File: 93. Restore IP Addresses/Solution.py
class Solution:
    # @param {string} s
    # @return {string[]}
    def restoreIpAddresses(self, s):
        if s == None or len(s) < 4 or len(s) > 12:
            return []
            
        res = []
        self.dfs(s, "", res, 0);
        return res
        
    def dfs(self, s, tmp, res, count):
        if count == 3 and self.isValid(s):
            res.append(tmp + s)
            return
        
        i = 1
        while i < 4 and i < len(s):
            substr = s[0:i]
            if self.isValid(substr):
                self.dfs(s[i:], tmp + substr + ".", res, count + 1)
            i += 1
    
    def isValid(self, s):
        if s == None or len(s) < 0:
            return False
        if s[0] == '0':
            return s == "0"
        return int(s) > 0 and int(s) <= 255
            
            
        

File: 93. Restore IP Addresses/Solution.java
public class Solution {
    public List<String> restoreIpAddresses(String s) {
        List<String> res = new ArrayList<String>();
        if (s == null || s.length() < 4 || s.length() > 12)
            return res;
            
        dfs(s, "", res, 0);
        return res;
    }
    
    private void dfs(String s, String tmp, List<String> res, int count) {
        if (count == 3 && isValid(s)) {
            res.add(tmp + s);
            return;
        }
        
        for (int i = 1; i < 4 && i < s.length(); i++) {
            String substr = s.substring(0, i);
            if (isValid(substr))
                dfs(s.substring(i), tmp + substr + ".", res, count+1);
        }
    }
    
    private boolean isValid(String s) {
        if (s == null || s.length() == 0)
            return false;
        if (s.charAt(0) == '0')
            return s.equals("0");
        int num = Integer.valueOf(s);
        return num <= 255 && num > 0;
    }
}